open-craft/edx-platform
# MIT CAPA: Withhold score if show_correctness condition is not met

Uses `correctness_available()` to withhold correct/incorrect feedback from the learner.

* Withholds the current score details from the learner by sending a null/undefined current problem score via AJAX and in the HTML templates.
* Replaces the previous correct/incorrect message with the Studio-configurable `submitted_message` (default is "Answer received.")
* Replaces the previous `correct`/`incorrect` status string with `answered`, which affects the icons, colors, and tooltip messages displayed.
* Changes the problem progress text to this text, both before and after an answer is submitted:
    `x point(s) possible (graded, withheld)`
    `x point(s) possible (ungraded, withheld)`

**JIRA tickets**: OSPR

**Discussions**: [Technical discovery document](https://docs.google.com/document/d/1yGAbIl5lZUXB_BTFDp5dx75fEypQHA7Ipg7M9PATKaw/edit)

**Screenshots**:

![screen shot 2017-03-24 at 9 22 09 am](https://cloud.githubusercontent.com/assets/7556571/24273670/74cd1fb4-1073-11e7-822a-c27988e72cb6.png)
![screen shot 2017-03-22 at 7 17 37 pm](https://cloud.githubusercontent.com/assets/7556571/24195616/5108ad5c-0f4a-11e7-8f52-aeb7d9dd2890.png)
![screen shot 2017-03-22 at 7 18 41 pm](https://cloud.githubusercontent.com/assets/7556571/24195599/450e9174-0f4a-11e7-8e61-3b9d5bb0dbd6.png)
![screen shot 2017-03-22 at 7 19 44 pm](https://cloud.githubusercontent.com/assets/7556571/24195606/48c7d7da-0f4a-11e7-9fe4-6a55670fed15.png)

**Sandbox URL**:

* LMS: https://pr14737.sandbox.opencraft.hosting/
* Studio: https://studio-pr14737.sandbox.opencraft.hosting/

**Partner information**: 3rd party-hosted open edX instance

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: 1 April 2017

**Testing instructions**:

*Note*: the sandbox has a [MIT CAPA Improvements](https://pr14737.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA1+course), which contains examples using all the different "Show Correctness" settings, with graded and ungraded problems.

1. Add CAPA problem to unit in Studio.
1. Click "EDIT", then click "SETTINGS".
1.  Scroll down to the "Show Correctness" setting, and select:
   * "Always" to see the current, default behaviour.
   * "Never" to ensure that correctness is never shown, either to the learner or staff.
   * "Past Due" to ensure that correctness is not shown until after the section due date.
   * "Closed" to ensure that correctness is not shown unless all attempts have been used, or the due date (and grace period) have expired.
1. Save the change.
1. View the problem in Preview or the LMS.
1. Observe that (under the conditions where correctness is withheld), the "problem progress" text reads "N points possible (graded, withheld)" for graded sections, or "N points possible (ungraded, withheld)" for ungraded sections.
1. Submit correct or incorrect responses. 
1. Observe that (under the conditions where correctness is withheld):
  * The "problem progress" text is unchanged.
  * The text displayed is "Answer received"
  * The icon displayed for the submitted answer is an asterisk, not a check or X mark.
  * The outline colour (e.g. for numerical input, radio and checkbox problems) is now blue instead of green/red.

To test that correctness is shown when the conditions for withholding correctness have elapsed:
1. Re-visit a problem with a previous response, and observe that the correctness markers are now in place.
1. Submit a new answer to a problem, and observe that correctness is shown as normal.

To test the new `Submitted Message` setting:

1. Add CAPA problem to unit in Studio.
1. Click "EDIT", then click "SETTINGS".
1. Observe that:
   * "Submitted Message" setting is present.
   * Default value is "Answer received."
1. Enter a value that differs from the default, then click "Save".
1. View the problem in Preview mode.
1. Submit an answer to the problem. 
1. Observe that the configured `Submitted Message` appears where "Correct/Incorrect (x/y)" used to appear.

**Author notes and concerns**:

1. This PR also suppresses all hints/feedback given when an answer is submitted, because it was hard to know whether some aspect of correctness would leak through in those messages.
   However if this assumption is incorrect, I can reinstate them.
1. The Submitted Message displays in addition to the "notification" text which appears below the problem.   In a subsequent PR, this text will replace the notification text so that only a single message displays, and the notification's icon and colour will be addressed.
1. I chose to use the `fa-asterisk` icon to mark "answered" answers where correctness is withheld, which is also used to mark answers that are "partially correct" (though the colour in that case is green).
  Am open to suggestions for an alternate icon.
1. The "Show Correctness" problem setting isn't persisted to the course export.  If this is easy to fix, I'm happy to do it here; otherwise it will be addressed by a subsequent PR.
1. This PR maintains backward-compatibility with the existing, unsupported `show_correctness` and `submitted_message` XML properties.  
  But ultimately, I think we should update these templates to remove the special case code:
  * [`choicetext.html`](https://github.com/edx/edx-platform/blob/master/common/lib/capa/capa/templates/choicetext.html)
  * [`choicegroup.html`](https://github.com/edx/edx-platform/blob/master/common/lib/capa/capa/templates/choicegroup.html)
  * [`status_span.html`](https://github.com/edx/edx-platform/blob/master/common/lib/capa/capa/templates/status_span.html)

**Reviewers**
